src/java/org/apache/cassandra/service/AbstractWriteResponseHandler.java

```diff
@@ -58,8 +72,28 @@
     private final long queryStartNanoTime;
     private volatile boolean supportsBackPressure = true;
 
+    static class SpeculationContext
+    {
+        final ReplicaList initial;
+        final ReplicaList backups;
+        final IMutation mutation;
+        final StorageProxy.WritePerformer performer;
+        final String localDC;
+
+        public SpeculationContext(ReplicaList initial, ReplicaList backups, IMutation mutation, StorageProxy.WritePerformer performer, String localDC)
+        {
+            this.initial = initial;
+            this.backups = backups;
+            this.mutation = mutation;
+            this.performer = performer;
+            this.localDC = localDC;
+        }
+    }
+
+    private volatile SpeculationContext speculationContext = null;
```

Comment:
Doesn't need to be volatile since it is set before the write response handler is published to other threads.